server/svix-server/src/test_util.rs

```diff
@@ -0,0 +1,232 @@
+use std::{collections::HashMap, net::TcpListener, str::FromStr};
+
+use anyhow::{Context, Result};
+use reqwest::{
+    header::{HeaderMap, HeaderName, HeaderValue},
+    Client, StatusCode,
+};
+use serde::{de::DeserializeOwned, Serialize};
+
+pub struct TestClient {
+    base_uri: String,
+    auth_header: String,
+    client: Client,
+}
+
+pub enum Method {
+    Get,
+    Post,
+    Put,
+    Delete,
+}
+
+impl TestClient {
+    pub fn new(base_uri: String, auth_token: &str) -> TestClient {
+        TestClient {
+            base_uri,
+            auth_header: format!("Bearer {}", auth_token),
+            client: Client::new(),
+        }
+    }
+
+    pub async fn request<I: Serialize, O: DeserializeOwned>(
```

Comment:
As I said on Slack, I think this can be much simpler without all of the optionals and the base request class. It'll provide better type safety and much shorter code. Here's what I had in mind (should work, but haven't tested):
```rust
pub struct TestClient {
    base_uri: String,
    auth_header: String,
    client: Client,
}

impl TestClient {
    pub fn new(base_uri: String, auth_token: &str) -> TestClient {
        TestClient {
            base_uri,
            auth_header: format!("Bearer {}", auth_token),
            client: Client::new(),
        }
    }

    fn build_url(&self, path: &str) -> String {
        format!("{}/{}", self.base_uri, path)
    }

    fn add_headers(&self, request: RequestBuilder) -> RequestBuilder {
        request.header("Authorization", &self.auth_header)
    }

    pub async fn post<I: Serialize, O: DeserializeOwned>(
        &self,
        path: &str,
        input: I,
        status_code: StatusCode,
    ) -> std::result::Result<O, String> {
        let req = self.client.post(self.build_url(path));
        let req = self.add_headers(req).json(&input);
        let resp = req.send().await.unwrap();
        assert!(resp.status() == status_code);
        Ok(resp.json().await.unwrap())
    }

    pub async fn put<I: Serialize, O: DeserializeOwned>(
        &self,
        path: &str,
        input: I,
        status_code: StatusCode,
    ) -> std::result::Result<O, String> {
        let req = self.client.put(self.build_url(path));
        let req = self.add_headers(req).json(&input);
        let resp = req.send().await.unwrap();
        assert!(resp.status() == status_code);
        Ok(resp.json().await.unwrap())
    }

    pub async fn get<O: DeserializeOwned + Clone>(
        &self,
        path: &str,
        status_code: StatusCode,
    ) -> std::result::Result<O, String> {
        let req = self.client.get(self.build_url(path));
        let req = self.add_headers(req);
        let resp = req.send().await.unwrap();
        assert!(resp.status() == status_code);
        Ok(resp.json().await.unwrap())
    }

    pub async fn delete<O: DeserializeOwned + Clone>(
        &self,
        path: &str,
        status_code: StatusCode,
    ) -> std::result::Result<O, String> {
        let req = self.client.delete(self.build_url(path));
        let req = self.add_headers(req);
        let resp = req.send().await.unwrap();
        assert!(resp.status() == status_code);
        Ok(resp.json().await.unwrap())
    }
}
```
Some points:
- We don't need to return the status_code because we always assert that it's what we expect (what we want in the context of a test suite).
- Body is not optional in post/put - if you sent to send an empty Body send an empty serialize struct (we have one in utils called `EmptyResponse`).
- Response is also not optional, if you want to parse nothing try to parse into  `EmptyResponse`.
- We don't want to have the test client have logic about comparing the body because that's very specific, and anyway we want the body out because we sometimes do different kinds of comparisons (you'll see as you add more tests).
- No need for callbacks/closures, we just return the body so we can do whatever comparisons we want.
- I made these functions return a result and was too lazy to add an error type/`anyhow`. Though to be honest, they should probably just return `O` and always unwrap inside. We just assume everything works, and if it doesn't we want to throw anyway. No need to propagate the errors.

I removed the support for custom headers for now, but the way to do it is probably by having *state* on the client and just letting people change the headers associated with the client before making a request. Yeah, it's a bit ugly, but it doesn't really matter and it's much simpler.

Reply:
The advantage to leaving the `unwrap` in the test body instead of in a helper function is the panic message includes a file/line number in the test. Otherwise I see what you mean now :).

Reply:
Good point. I guess you can make it return a result and unwrap in the test!